### ISSUE TITLE
integration tests failing with "(-32000) Service connection disposed"

### ISSUE THREAD

**Issue by christopherfujino** [flutter/flutter#155475]
I believe that this is the root cause of https://github.com/flutter/flutter/issues/153578, which was that MANY integration tests in the flutter/packages repo were timing out. After I landed https://github.com/flutter/flutter/pull/154970, these should no longer time out, and I found https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8736798859005867345/+/u/Run_package_tests/drive_examples/stdout, which from the stacktrace would have timed out before my change.

```
00:39 +16 ~1 -1: loading /Volumes/Work/s/w/ir/x/w/packages/packages/webview_flutter/webview_flutter_wkwebview/example/integration_test/webview_flutter_test.dart [E]                                   
  ext.flutter.integrationTest: (-32000) Service connection disposed
  package:vm_service/src/vm_service.dart 261:34                      new _OutstandingRequest
  package:vm_service/src/vm_service.dart 1778:25                     VmService._call.<fn>
  package:vm_service/src/vm_service.dart 1790:8                      VmService._call
  package:vm_service/src/vm_service.dart 1729:14                     VmService.callServiceExtension
  package:flutter_tools/src/test/integration_test_device.dart 98:25  IntegrationTestTestDevice.start.<fn>
  dart:async/zone.dart 1422:47                                       _rootRunUnary
  dart:async/zone.dart 1324:19                                       _CustomZone.runUnary
  dart:async/zone.dart 1233:7                                        _CustomZone.runUnaryGuarded
  dart:async/stream_impl.dart 366:11                                 _BufferingStreamSubscription._sendData
  dart:async/stream_impl.dart 297:7                                  _BufferingStreamSubscription._add
  dart:async/stream_controller.dart 777:19                           _SyncStreamControllerDispatch._sendData
  dart:async/stream_controller.dart 651:7                            _StreamController._add
  dart:async/stream_controller.dart 606:5                            _StreamController.add
  dart:async/zone.dart 1422:47                                       _rootRunUnary
  dart:async/zone.dart 1324:19                                       _CustomZone.runUnary
  dart:async/zone.dart 1233:7                                        _CustomZone.runUnaryGuarded
  dart:async/stream_impl.dart 366:11                                 _BufferingStreamSubscription._sendData
  dart:async/stream_impl.dart 542:14                                 _DelayedData.perform
  dart:async/stream_impl.dart 647:11                                 _PendingEvents.handleNext
  dart:async/stream_impl.dart 618:7                                  _PendingEvents.schedule.<fn>
  dart:async/zone.dart 1406:47                                       _rootRun
  dart:async/zone.dart 1317:19                                       _CustomZone.run
  dart:async/zone.dart 1225:7                                        _CustomZone.runGuarded
  dart:async/zone.dart 1265:23                                       _CustomZone.bindCallbackGuarded.<fn>
  dart:async/zone.dart 1414:13                                       _rootRun
  dart:async/zone.dart 1317:19                                       _CustomZone.run
  dart:async/zone.dart 1225:7                                        _CustomZone.runGuarded
  dart:async/zone.dart 1265:23                                       _CustomZone.bindCallbackGuarded.<fn>
  dart:async/schedule_microtask.dart 40:21                           _microtaskLoop
  dart:async/schedule_microtask.dart 49:5                            _startMicrotaskLoop
  dart:isolate-patch/isolate_patch.dart 118:13                       _runPendingImmediateCallback
  dart:isolate-patch/timer_impl.dart 405:11                          _Timer._runTimers
  dart:isolate-patch/timer_impl.dart 429:5                           _Timer._handleMessage
  dart:isolate-patch/isolate_patch.dart 184:12                       _RawReceivePort._handleMessage
```

I'm also seeing this in a lot of devicelab tests in the flutter/flutter repo:

- https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8736499130357881889/+/u/run_new_gallery__transition_perf/stdout
- https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8736262697040472033/+/u/run_gradle_plugin_fat_apk_test/stdout
- https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8736260054150681217/+/u/run_draw_vertices_perf_opengles__timeline_summary/stdout

**Comment by a-siva** [comment]
//cc @derekxu16 @bkonyi

**Comment by jtmcdole** [comment]
Similar failure on Linux pixel 7pro test (to the closed bug):

```
[2024-09-25 09:03:30.302097] [STDOUT] Executing "git checkout /opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery/android/app/src/main/AndroidManifest.xml" in "/opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery" with environment {BOT: true, LANG: en_US.UTF-8}
[2024-09-25 09:03:30.344436] [STDOUT] stderr: error: pathspec '/opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery/android/app/src/main/AndroidManifest.xml' did not match any file(s) known to git
[2024-09-25 09:03:30.349596] [STDOUT] Caught exception while trying to reset AndroidManifest: Executable "git" failed with exit code 1.
[2024-09-25 09:03:30.350486] [STDOUT] Executing "git checkout /opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery/ios/Runner/Info.plist" in "/opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery" with environment {BOT: true, LANG: en_US.UTF-8}
[2024-09-25 09:03:30.373469] [STDOUT] stderr: error: pathspec '/opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery/ios/Runner/Info.plist' did not match any file(s) known to git
[2024-09-25 09:03:30.375029] [STDOUT] Caught exception while trying to reset Info.plist: Executable "git" failed with exit code 1.
[2024-09-25 09:03:30.376115] [STDOUT] Executing "/opt/s/w/ir/cache/android/sdk/platform-tools/adb -s 33271FDH3002BW shell cmd power set-fixed-performance-mode-enabled false" in "/opt/s/w/ir/x/w/rc/tmp931g0asa/flutter sdk/dev/integration_tests/new_gallery" with environment {BOT: true, LANG: en_US.UTF-8}
[2024-09-25 09:03:30.437287] [STDOUT] Checking for reboot
[new_gallery_impeller_old_zoom__transition_perf] Task runner system failed with exception!
ext.cocoonRunTask: (-32000) Service connection disposed
#0      new _OutstandingRequest (package:vm_service/src/vm_service.dart:261:34)
#1      VmService._call.<anonymous closure> (package:vm_service/src/vm_service.dart:1778:25)
#2      VmService._call (package:vm_service/src/vm_service.dart:1790:8)
#3      VmService.callServiceExtension (package:vm_service/src/vm_service.dart:1729:14)
#4      runTask (package:flutter_devicelab/framework/runner.dart:240:73)
<asynchronous suspension>
#5      rerunTask (package:flutter_devicelab/framework/runner.dart:119:29)
<asynchronous suspension>
#6      runTasks (package:flutter_devicelab/framework/runner.dart:52:16)
<asynchronous suspension>
#7      TestCommand.run (package:flutter_devicelab/command/test.dart:86:5)
<asynchronous suspension>
#8      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
```

https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20new_gallery_impeller_old_zoom__transition_perf/4465/overview

These are getting marked as test flakes.

**Comment by bkonyi** [comment]
This is likely due to the target application exiting before the RPC is completed. It might be enough to just catch the `RPCError`, check for the disposed error code, and then wait on the process exit code future with a short timeout to ensure the process is actually exiting properly.

**Comment by christopherfujino** [comment]
> This is likely due to the target application exiting before the RPC is completed. It might be enough to just catch the `RPCError`, check for the disposed error code, and then wait on the process exit code future with a short timeout to ensure the process is actually exiting properly.

Do we need to differentiate between exited successfully and exited with a failure? Or I guess they would be handled if we're checking the exit code?

**Comment by bkonyi** [comment]
Yeah, we'd still check the exit code to determine if we exited successfully or not. However, after giving this some more thought, `runTask` returns the `TaskResult` which is the result of the RPC that's failing to complete. That seems to imply that the target is exiting before actually responding to the `ext.cocoonRunTask` RPC which sounds like a bug.

**Comment by jtmcdole** [comment]
Do we have an easy way to check and mark things as infra problems? Right now we've seen these show up as "test flakes", which creates extra work.

**Comment by jmagman** [comment]
> Do we have an easy way to check and mark things as infra problems? Right now we've seen these show up as "test flakes", which creates extra work.

It's not definitively an infra issue if the app is crashing... Hopefully we'll know more if we have the exit code or some better stack from the crashing app.

**Comment by bkonyi** [comment]
So it looks like the `_TaskRunner` in the task waits for the runner process to invoke `ext.cocoonRunTask`, which waits for the task to execute before returning the `TaskResult`. However, `_TaskRunner.run(...)` closes the ports keeping the isolate around before returning, so once the callback for the service extension completes the main isolate is able to exit. This causes the VM to shutdown while the service isolate is still processing the service extension response.

I think the simplest solution here is to add another service extension (e.g., `ext.cocoonTaskResultsReceived`) that the runner process can invoke to notify the task that it can shutdown.

**Comment by jmagman** [comment]
Something similar is happening in the packages repo with the non-Flutter devicelab totally bespoke way of doing this:   `ext.flutter.integrationTest: (-32000) Service connection disposed`.  https://github.com/flutter/flutter/issues/153578 @bkonyi are we just working around this by fixing it only in the devicelab runner?

**Comment by bkonyi** [comment]
> Something similar is happening in the packages repo with the non-Flutter devicelab totally bespoke way of doing this: `ext.flutter.integrationTest: (-32000) Service connection disposed`. #153578 @bkonyi are we just working around this by fixing it only in the devicelab runner?

We're going to need to perform a similar fix there. Ideally we'd have a way for the VM service to keep the VM alive while it's still processing responses, but that would be a much bigger project.

**Comment by Renzo-Olivares** [comment]
Seeing a similar failure the last few times the tree has gone down in the last hour, `Linux_pixel_7pro new_gallery_impeller_old_zoom__transition_perf` : https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_pixel_7pro%20new_gallery_impeller_old_zoom__transition_perf/4547/overview

```
[new_gallery_impeller_old_zoom__transition_perf] Task runner system failed with exception!
ext.cocoonRunTask: (-32000) Service connection disposed
#0      new _OutstandingRequest (package:vm_service/src/vm_service.dart:265:34)
#1      VmService._call.<anonymous closure> (package:vm_service/src/vm_service.dart:1921:25)
#2      VmService._call (package:vm_service/src/vm_service.dart:1933:8)
#3      VmService.callServiceExtension (package:vm_service/src/vm_service.dart:1872:14)
#4      runTask (package:flutter_devicelab/framework/runner.dart:240:73)
<asynchronous suspension>
#5      rerunTask (package:flutter_devicelab/framework/runner.dart:119:29)
<asynchronous suspension>
#6      runTasks (package:flutter_devicelab/framework/runner.dart:52:16)
<asynchronous suspension>
#7      TestCommand.run (package:flutter_devicelab/command/test.dart:86:5)
<asynchronous suspension>
#8      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>

[new_gallery_impeller_old_zoom__transition_perf] Terminating process...
ext.cocoonRunTask: (-32000) Service connection disposed
```

**Comment by bkonyi** [comment]
> > We're going to need to perform a similar fix there.
> 
> In the integration_test package? Does that need another issue, or are you tracking it with this one?

We can just track in https://github.com/flutter/flutter/issues/153578. I'll take a look at that issue once this one is resolved.

> > Ideally we'd have a way for the VM service to keep the VM alive while it's still processing responses, but that would be a much bigger project.
> 
> Makes sense! Do you have an issue for that project?

I've filed https://github.com/dart-lang/sdk/issues/56831 to track this, although I don't expect we'll get around to this in the near future.
